### PR TITLE
Add vulnerability search API

### PR DESCRIPTION
**Pull request by stonezdj**
use q.Query to pass all query conditions

Signed-off-by: stonezdj <daojunz@vmware.com>

Thank you for contributing to Harbor!

# Comprehensive Summary of your change

# Issue being fixed
Fixes #(issue)

Please indicate you've done the following:
- [x] Well Written Title and Summary of the PR
- [x] Label the PR as needed. "release-note/ignore-for-release, release-note/new-feature, release-note/update, release-note/enhancement, release-note/community, release-note/breaking-change, release-note/docs, release-note/infra, release-note/deprecation"
- [x] Accepted the DCO. Commits without the DCO will delay acceptance.
- [x] Made sure tests are passing and test coverage is added if needed.
- [x] Considered the docs impact and opened a new docs issue or PR with docs changes if needed in [website repository](https://github.com/goharbor/website).

### CODE REVIEW

File: api/v2.0/swagger.yaml [hunk 1]
```diff
@@ -6086,6 +6086,55 @@ paths:
...
+        and tag(exact match) 
+      tags:
+        - securityhub
+      operationId: ListVulnerabilities
+      parameters:
+        - $ref: '#/parameters/requestId'
+        - $ref: '#/parameters/query'
+        - $ref: '#/parameters/page'
+        - $ref: '#/parameters/pageSize'
+        - name: tune_count
```
Comment (chlins) [inline api/v2.0/swagger.yaml:6111]: `total_count`?

File: src/server/v2.0/handler/security.go [hunk 2]
```diff
@@ -96,3 +97,52 @@ func toDangerousCves(cves []*scan.VulnerabilityRecord) []*models.DangerousCVE {
 	}
 	return result
 }
+
+func (s *securityAPI) ListVulnerabilities(ctx context.Context, params securityModel.ListVulnerabilitiesParams) middleware.Responder {
+	if err := s.RequireSystemAccess(ctx, rbac.ActionRead, rbac.ResourceSecurityHub); err != nil {
```
Comment (chlins) [inline src/server/v2.0/handler/security.go:102]: `rbac.ActionList`?

File: src/pkg/securityhub/dao/security.go [hunk 3]
```diff
@@ -131,3 +231,118 @@ func (d *dao) DangerousCVEs(ctx context.Context, scannerUUID string, projectID i
...
+	err = o.Raw(countSQL(sqlStr), params).QueryRow(&cnt)
+	return cnt, err
+}
+
+func (d *dao) countExceedLimit(ctx context.Context, sqlStr string, params []interface{}) (bool, error) {
+	o, err := orm.FromContext(ctx)
+	if err != nil {
+		return true, err
+	}
+	queryExceed := fmt.Sprintf(`SELECT EXISTS (%v LIMIT 1 OFFSET 1000)`, sqlStr)
```
Comment (chlins) [inline src/pkg/securityhub/dao/security.go:270]: `%v` => `%s`

File: src/pkg/securityhub/dao/security.go [hunk 4]
```diff
@@ -131,3 +231,118 @@ func (d *dao) DangerousCVEs(ctx context.Context, scannerUUID string, projectID i
...
+	if err := checkQFilter(query, allowedFieldTypeMap); err != nil {
+		return 0, err
+	}
+	sqlStr, params = applyVulFilter(ctx, sqlStr, query, params)
+	if tuneCount {
+		exceedLimit, err := d.countExceedLimit(ctx, sqlStr, params)
+		if err != nil {
+			return 0, err
+		}
+		log.Infof("exceedLimit: %v", exceedLimit)
```
Comment (chlins) [inline src/pkg/securityhub/dao/security.go:255]: I think `Warnning` is better, and log should be clear like `The count is exceed to limit %v due to the tuneCount is enabled`.

File: src/controller/securityhub/controller.go [hunk 5]
```diff
@@ -63,12 +64,19 @@ func WithArtifact(enable bool) Option {
 type Controller interface {
 	// SecuritySummary returns the security summary of the specified project.
 	SecuritySummary(ctx context.Context, projectID int64, options ...Option) (*secHubModel.Summary, error)
+	// ListVuls list vulnerabilities by query
+	ListVuls(ctx context.Context, scannerUUID string, projectID int64, query *q.Query) ([]*secHubModel.VulnerabilityItem, error)
+	// CountVuls get all vulnerability count by query
+	CountVuls(ctx context.Context, scannerUUID string, projectID int64, tuneCount bool, query *q.Query) (int64, error)
+	// DefaultScannerUUID get default scanner UUID
+	DefaultScannerUUID(ctx context.Context) (string, error)
```
Comment (wy65701436) [inline src/controller/securityhub/controller.go:72]: we should put the DefaultScannerUUID in the scannerMgr

Comment (stonezdj) [reply src/controller/securityhub/controller.go]: Updated

File: src/controller/securityhub/controller.go [hunk 6]
```diff
@@ -128,11 +135,50 @@ func (c *controller) totalArtifactCount(ctx context.Context, projectID int64) (i
...
+		return nil, err
+	}
+	resultList, err := c.attachTags(ctx, vuls)
+	if err != nil {
+		return nil, err
+	}
+	return resultList, nil
+}
+
+func (c *controller) attachTags(ctx context.Context, vuls []*secHubModel.VulnerabilityItem) ([]*secHubModel.VulnerabilityItem, error) {
```
Comment (wy65701436) [inline src/controller/securityhub/controller.go:150]: add a with_tag parameter?

File: src/pkg/securityhub/dao/security.go [hunk 7]
```diff
@@ -131,3 +215,119 @@ func (d *dao) DangerousCVEs(ctx context.Context, scannerUUID string, projectID i
...
+	queryExceed := fmt.Sprintf(`SELECT EXISTS (%s LIMIT 1 OFFSET 1000)`, sqlStr)
+	var exceed bool
+	err = o.Raw(queryExceed, params).QueryRow(&exceed)
+	if err != nil {
+		return true, err
+	}
+	return exceed, nil
+}
+
+func (d *dao) ListVulnerabilities(ctx context.Context, registrationUUID string, projectID int64, query *q.Query) ([]*model.VulnerabilityItem, error) {
```
Comment (wy65701436) [inline src/pkg/securityhub/dao/security.go:263]: if query == nil {
		return
}

File: src/pkg/securityhub/dao/security.go [hunk 8]
```diff
@@ -131,3 +215,119 @@ func (d *dao) DangerousCVEs(ctx context.Context, scannerUUID string, projectID i
...
+	newParam = params
+	for k, m := range filterMap {
+		s, p := m.FilterFunc(ctx, k, query)
+		queryStr = queryStr + s
+		newParam = append(newParam, p...)
+	}
+	return queryStr, newParam
+}
+
+func applyVulPagination(sqlStr string, query *q.Query, params []interface{}) (string, []interface{}) {
```
Comment (wy65701436) [inline src/pkg/securityhub/dao/security.go:291]: please add comments for the hidden logic, like order by cvss_score_v3 desc

File: src/pkg/securityhub/dao/security.go [hunk 9]
```diff
@@ -131,3 +215,119 @@ func (d *dao) DangerousCVEs(ctx context.Context, scannerUUID string, projectID i
...
+			log.Warning("the count is exceed to limit 1000 due to the tuneCount is enabled, return count with -1 instead")
+			return -1, nil
+		}
+	}
+	var cnt int64
+	err = o.Raw(countSQL(sqlStr), params).QueryRow(&cnt)
+	return cnt, err
+}
+
+func (d *dao) countExceedLimit(ctx context.Context, sqlStr string, params []interface{}) (bool, error) {
```
Comment (wy65701436) [inline src/pkg/securityhub/dao/security.go:249]: add some comment to describe in which situation the result is true or false.

**Review by wy65701436 — APPROVED** [review]
lgtm

File: src/pkg/securityhub/dao/security.go [hunk 10]
```diff
@@ -131,3 +215,119 @@ func (d *dao) DangerousCVEs(ctx context.Context, scannerUUID string, projectID i
...
+	o, err := orm.FromContext(ctx)
+	if err != nil {
```
Comment (wy65701436) [inline src/pkg/securityhub/dao/security.go:251]: return false?

File: src/pkg/securityhub/dao/security.go [hunk 11]
```diff
@@ -131,3 +215,119 @@ func (d *dao) DangerousCVEs(ctx context.Context, scannerUUID string, projectID i
 	_, err = o.Raw(dangerousCVESQL, scannerUUID).QueryRows(&cves)
 	return cves, err
 }
+
+func countSQL(strSQL string) string {
+	return fmt.Sprintf(`select count(1) cnt from (%v) as t`, strSQL)
+}
+
+func (d *dao) CountVulnerabilities(ctx context.Context, registrationUUID string, projectID int64, tuneCount bool, query *q.Query) (int64, error) {
```
[the remark this answers is not in the document]
Comment (stonezdj) [reply src/pkg/securityhub/dao/security.go:223]: can't return directly

File: src/pkg/securityhub/dao/security.go [hunk 12]
```diff
@@ -131,3 +215,119 @@ func (d *dao) DangerousCVEs(ctx context.Context, scannerUUID string, projectID i
...
+		return nil, err
+	}
+	sqlStr, params = applyVulFilter(ctx, sqlStr, query, params)
+	sqlStr, params = applyVulPagination(sqlStr, query, params)
+	vulnRecs := make([]*model.VulnerabilityItem, 0)
+	_, err = o.Raw(sqlStr, params).QueryRows(&vulnRecs)
+	return vulnRecs, err
+}
+
+func applyVulFilter(ctx context.Context, sqlStr string, query *q.Query, params []interface{}) (queryStr string, newParam []interface{}) {
```
[the remark this answers is not in the document]
Comment (stonezdj) [reply src/pkg/securityhub/dao/security.go:280]: Updated

**Review by chlins — APPROVED** [review]
lgtm

**Pull request merged**